### DISCUSSION TITLE
How to open Netbeans project in Netbeans?

### DISCUSSION THREAD

**Discussion by bykka** [apache/netbeans#5097] [Q&A]
Hi,
I have checked out the source code of Netbeans and would like to open it in Netbeans 16 for running and debugging.
How to do that? Because Netbeans does not see it as a project.

**Comment by errael** [comment]
After downloading source, I'll first build it. See https://netbeans.apache.org/download/nb16/

And there's a section `Learn to run and debug NetBeans IDE or Platform applications` at https://netbeans.apache.org/participate/submit-pr.html which you may find useful.

Basically, register the platform you built, create a dummy nbm project and use the platform you built/registered, open the NetBeans module you're interested in (for example `source-location/platform/api.io`, set a breakpoint, debug the dummy module.

You can't open all the NB modules with one command.

**Comment by bykka** [comment]
With Ant I can build everything and run. Pity that it is not possible to do the same with Netbeans because expected to use it for finding usages(dependencies) instead of "grepping".

**Comment by neilcsmith-net** [comment]
I can't say I've done most of the steps on @errael link, and haven't registered the IDE as a platform or created a dummy NBM project.  It is worth adding it in the Favorites window.

I would recommend building the IDE from the command line first, but you should be able to build it by finding the root folder in the Favorites window, right-clicking on `build.xml` and choosing `build`.  Just make sure you're running your IDE on JDK 11 if you do this.  If you're running the IDE on a later JDK you'll have to set some additional properties to control the JDK used for building.

For the project you want to look at, find the folder in the Favorites window, right-click and choose `Open Project of Folder`.  You can run and debug the IDE from any project you have open.

Unfortunately this does miss the ability to find usages across the whole IDE, but most of the time the open projects and dependencies scope provides the info I want, and if not there's always `CTRL-F` on the IDE root folder in the Favorites window or `ag` on the CLI. :shrug:

**Comment by bykka** [comment]
Thank you guys, I will try.

Just want to emphasize that process is not intuitive (obvious) and I guess discourages new developers to contribute.

**Reply by mbien** [reply]
some of the doc might be a bit outdated, but its basically three steps:
 - build NB once via command line
 - open the modules you are interested in like a regular project in NetBeans
 - now you can run / debug / test / code from there

clean build everything again if you check out new changes from master

you don't want to open all 845 projects since it is going to slow everything down. However, I have done it before for running project wide code inspections, it is possible just not necessary or recommended for most situations.

How to find the right module? Well that is often a challenge. Sometimes running a full text search for a window title or menu entry is all it takes.

**Reply by mbien** [reply]
going to update the readme via #5126. I hope this clarifies things. If something is missing, let me know.

**Comment by errael** [comment]
> haven't registered the IDE as a platform or created a dummy NBM project

I'm frequently working with a plugin (not just any dummy project) which is where these steps come from. And I'll often times use an actually dummy project that uses some NB module development feature; then I can put breakpoints in the dummy module and see how it got there from NetBeans. BTW, If you have a nbm/dummy-module open, right click on it, select `New > Other`, from `Categories: Module Development` there are a variety of things you can put into the dummy module. For example, with `Action` you can put something in a Menu or Toolbar (or both). Click run on the test module and you'll get an instance of NB with your menu item.

@neilcsmith-net thanks for the hint about how favorites/build.xml can be used; I'd forgotten about that stuff.

**Reply by neilcsmith-net** [reply]
Yes, I can definitely see the benefit.  Just think those instructions are a little more complicated than they need to be for new developers.

**Comment by Chris2011** [comment]
NetBeans is module based. The whole NetBeans source is not a project at all. You need to open the module, that you want to work on. If you work on PHP then navigate to the correct subfolder and you will see that there are a lot of projects.
